service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java

```diff
@@ -77,21 +82,123 @@
+                        .filter(response -> response.getValue().getAssignee().getValue().equals(assignee))
+                        .map(Element::getValue)
+                        .findFirst()
+                        .orElse(null))
+                    .build())
+                .build())
+            .collect(toList());
+    }
+
+    /**
+     * Collects directions for all parties and places them into single map, where the key is the role direction,
+     * and the value is the list of directions associated.
+     *
+     * @param caseData data from case.
+     * @return Map of roles and directions.
+     */
+    public Map<DirectionAssignee, List<Element<Direction>>> collectDirectionsToMap(CaseData caseData) {
+        final List<Element<Direction>> emptyList = ImmutableList.of();
+
+        // courtDirectionsCustom is used due to the workaround in ComplyWithDirectionsController
+        return Map.of(
+            ALL_PARTIES, defaultIfNull(caseData.getAllParties(), emptyList),
+            LOCAL_AUTHORITY, defaultIfNull(caseData.getLocalAuthorityDirections(), emptyList),
+            CAFCASS, defaultIfNull(caseData.getCafcassDirections(), emptyList),
+            COURT, defaultIfNull(caseData.getCourtDirectionsCustom(), emptyList),
+            PARENTS_AND_RESPONDENTS, defaultIfNull(caseData.getParentsAndRespondentsDirections(), emptyList),
+            OTHERS, defaultIfNull(caseData.getOtherPartiesDirections(), emptyList));
+    }
+
+    /**
+     * Adds directionId and assignee wherever a response is present within a direction.
+     *
+     * @param directionMap a map of directions where assignee is key and value is a list of directions.
+     * @return a list of directions with hidden variables added where a response is present.
+     */
+    public List<Element<Direction>> addHiddenVariablesToResponseForManyAssignees(
+        Map<DirectionAssignee, List<Element<Direction>>> directionMap) {
+        return directionMap.entrySet()
+            .stream()
+            .map(entry -> addHiddenVariablesToResponseForAssignee(entry.getKey(), entry.getValue()))
+            .flatMap(List::stream)
+            .collect(toList());
+    }
+
+    private List<Element<Direction>> addHiddenVariablesToResponseForAssignee(DirectionAssignee assignee,
+                                                                             List<Element<Direction>> directions) {
+        return directions.stream()
+            .filter(element -> isNotEmpty(element.getValue().getResponse()))
+            .filter(element -> isNotEmpty(element.getValue().getResponse().getComplied()))
+            .map(x -> Element.<Direction>builder()
+                .id(x.getId())
+                .value(x.getValue().toBuilder()
+                    .response(x.getValue().getResponse().toBuilder()
+                        .assignee(assignee)
+                        .directionId(x.getId())
+                        .build())
+                    .build())
+                .build())
```

Comment:
That map probably could be extracted. Logic not relevant to what's happening here.

Reply:
I disagree, the map builds the response. I have moved the filter methods to a private method